# Дашборд для Яндекс.Дзен

### Описание проекта

Вы работаете аналитиком в Яндекс.Дзене. Почти всё ваше время занимает анализ пользовательского взаимодействия с карточками статей.

Каждую карточку определяют её тема и источник (у него тоже есть тема). Примеры тем: «Красота и здоровье», «Россия», «Путешествия». А пользователей системы характеризует возрастная категория.

Есть три способа взаимодействия пользователей с системой:

    Карточка отображена для пользователя (show);
    Пользователь кликнул на карточку (click);
    Пользователь просмотрел статью карточки (view).
    Каждую неделю начинающие менеджеры Денис и Валерия задают вам одни и те же вопросы:

    Сколько взаимодействий пользователей с карточками происходит в системе с разбивкой по темам карточек?
    Как много карточек генерируют источники с разными темами?
    Как соотносятся темы карточек и темы источников?

На шестую неделю работы вы решаете, что процесс пора автоматизировать. Для Дениса и Валерии нужно сделать дашборд.

Дашборд будет основываться на пайплайне, который будет брать данные из таблицы, в которых хранятся сырые данные, трансформировать данные и укладывать их в агрегирующую таблицу. Пайплайн будет разработан для вас дата-инженерами.
Краткое ТЗ для создания дашборда:

*Бизнес-задача: анализ взаимодействия пользователей с карточками Яндекс.Дзен; Насколько часто предполагается пользоваться дашбордом: не реже, чем раз в неделю; Кто будет основным пользователем дашборда*: менеджеры по анализу контента;

*Состав данных для дашборда:*

    История событий по темам карточек (два графика - абсолютные числа и процентное соотношение);
    Разбивка событий по темам источников;
    Таблица соответствия тем источников темам карточек;

*По каким параметрам данные должны группироваться:*

    Дата и время;
    Тема карточки;
    Тема источника;
    Возрастная группа;
    Характер данных:

История событий по темам карточек — абсолютные величины с разбивкой по минутам; Разбивка событий по темам источников — относительные величины (% событий); Соответствия тем источников темам карточек - абсолютные величины;

*Важность:* все графики имеют равную важность; Источники данных для дашборда:

Сырые данные о событиях взаимодействия пользователей с карточками (таблица dash_visits); База данных, в которой будут храниться агрегированные данные:

Дополнительные агрегированные таблицы в БД zen;

*Частота обновления данных:* Один раз в сутки, в полночь по UTC;
Задачи проекта

    Получить данные из базы данных

    Создать дашборд на основе полученных данных

    Ответить на вопросы менеджеров в виде презентации:

    Cколько взаимодействий пользователей с карточками происходит в системе с разбивкой по темам карточек?

    Как много карточек генерируют источники с разными темами?

    Как соотносятся темы карточек и темы источников?



In [1]:
# импорт библиотек
import pandas as pd
from sqlalchemy import create_engine
from IPython.display import Image

*Макет дашборда:*

In [2]:
Image(url='https://pictures.s3.yandex.net/resources/Untitled_-_2020-07-06T160925.436_1594041010.png')

In [3]:
# подключение
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-zen-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string) 

In [4]:
query = '''
           SELECT *
           FROM dash_visits
        '''
dash_visits = pd.io.sql.read_sql(query, con=engine)

In [5]:
dash_visits.head(5) 

record_id item_topic source_topic age_segment                  dt  visits
0    1040597     Деньги         Авто       18-25 2019-09-24 18:32:00       3
1    1040598     Деньги         Авто       18-25 2019-09-24 18:35:00       1
2    1040599     Деньги         Авто       18-25 2019-09-24 18:54:00       4
3    1040600     Деньги         Авто       18-25 2019-09-24 18:55:00      17
4    1040601     Деньги         Авто       18-25 2019-09-24 18:56:00      27

In [6]:
# проверка датафрейма
print(dash_visits.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30745 entries, 0 to 30744
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   record_id     30745 non-null  int64         
 1   item_topic    30745 non-null  object        
 2   source_topic  30745 non-null  object        
 3   age_segment   30745 non-null  object        
 4   dt            30745 non-null  datetime64[ns]
 5   visits        30745 non-null  int64         
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 1.4+ MB
None


In [7]:
print(dash_visits.describe())

          record_id        visits
count  3.074500e+04  30745.000000
mean   1.055969e+06     10.089673
std    8.875461e+03     19.727601
min    1.040597e+06      1.000000
25%    1.048283e+06      1.000000
50%    1.055969e+06      3.000000
75%    1.063655e+06     10.000000
max    1.071341e+06    371.000000


In [8]:
# проверка на явные дубликаты
print('Явных дубликатов:', dash_visits.duplicated().sum())

Явных дубликатов: 0


In [27]:
# самые популярные карточки по событиям
test1 = dash_visits.groupby(by = 'item_topic')['visits'].sum()
test1.sort_values(ascending=False)

item_topic
Наука                 21736
Отношения             20666
Интересные факты      19942
Общество              19640
Подборки              17772
Россия                16966
Полезные советы       15435
История               15389
Семья                 11897
Женщины               11499
Дети                  10989
Рассказы              10909
Здоровье              10399
Деньги                10291
Культура              10205
Красота                9814
Туризм                 9512
Юмор                   9398
Скандалы               9294
Путешествия            9260
Искусство              8516
Психология             8036
Женская психология     7737
Шоу                    7511
Знаменитости           7394
Name: visits, dtype: int64

In [28]:
test1.sum()

310207

In [9]:
# Типы карточек
dash_visits['item_topic'].value_counts()

Отношения             1536
Интересные факты      1535
Наука                 1505
Подборки              1456
Полезные советы       1424
Общество              1422
Россия                1385
История               1363
Семья                 1287
Путешествия           1247
Деньги                1234
Женщины               1230
Дети                  1229
Туризм                1206
Здоровье              1203
Красота               1193
Культура              1160
Юмор                  1129
Искусство             1119
Рассказы              1109
Психология            1056
Скандалы              1023
Знаменитости           976
Женская психология     914
Шоу                    804
Name: item_topic, dtype: int64

In [10]:
# Темы источников
dash_visits['source_topic'].value_counts()

Семейные отношения    1822
Россия                1687
Знаменитости          1650
Полезные советы       1578
Путешествия           1563
Кино                  1505
Дети                  1459
История               1437
Семья                 1405
Одежда                1379
Здоровье              1243
Искусство             1228
Авто                  1077
Психология            1055
Сад и дача            1036
Политика              1024
Спорт                 1007
Сделай сам             995
Ремонт                 985
Деньги                 973
Еда                    912
Интерьеры              809
Строительство          758
Музыка                 750
Технологии             741
Финансы                667
Name: source_topic, dtype: int64

In [16]:
# возрастные группы
dash_visits['age_segment'].value_counts()

18-25    7056
26-30    5875
31-35    5552
36-40    5105
41-45    3903
45+      3254
Name: age_segment, dtype: int64

In [12]:
# период
print(dash_visits['dt'].min(), '||', dash_visits['dt'].max())

2019-09-24 18:28:00 || 2019-09-24 19:00:00


In [13]:
# преобразование в csv
dash_visits.to_csv('dash_visits.csv', index = False)

Дашборд:

https://public.tableau.com/app/profile/dm.grig/viz/sptint16au/Result

Презентация:

https://disk.yandex.ru/d/eg-QgZ52RrVlLg

## Общий вывод:


В ходе аналитической работы получен датафрейм, содержащий 30745 записей по 5 столбцам, что соотутветсвует заданию.  Дубликаты и пропуски отсутствуют.

Построив дашборд мы можем отметить следующие посещения карточек:

На самый пик событий в 18.58 приходится суммарно 61.247 посещений.

*Топ-3 наиболее популярных тем карточек:*

    Наука (4372 посещения в пик, 21736 всего за весь период) 
    Отношения (4145 посещений в пик, 20666 всего за весь период)
    Интересные факты (3910 посещений в пик, 19942 всего за весь период)
    

*По теме источников генерируют следующие количества посещенений:*

Топ-3 наиболее популярных тем источников:

    Семейные отношения (10.7% переходов)
    Россия (9.6% переходов)
    Полезные советы (8.8% переходов)
    
*Топ-5 наиболее популярных связок «источник - тема карточки»:*

    Путешествия - Рассказы (4.587 переходов)
    Россия - Общество (3.471 переходов)
    Кино - Наука (3.279 переходов)
    
